RATapi/examples/bayes_benchmark/bayes_benchmark.py

```diff
@@ -0,0 +1,361 @@
+
+        Parameters
+        ----------
+        roughness_index : int
+            The index of the roughness vector to use as the roughness parameter value.
+        background_index : int
+            The index of the background vector to use as the background parameter value.
+        scalefactor_index : int
+            The index of the scalefactor vector to use as the scalefactor parameter.
+
+        Returns
+        -------
+        float
+            The value of exp(-chi^2 / 2) for the given roughness and background values.
+        """
+        problem.parameters[0].value = roughness[roughness_index]
+        problem.background_parameters[0].value = background[background_index]
+        problem.scalefactors[0].value = scalefactor[scalefactor_index]
+
+        _, results = RAT.run(problem, controls)
+        chi_squared = results.calculationResults.sumChi
+
+        return np.exp(-chi_squared / 2)
+
+    vectorized_calc_posterior = np.vectorize(calculate_posterior)
+
+    print("Calculating posterior directly...")
+    probability_array = vectorized_calc_posterior(*np.indices((grid_size, grid_size, grid_size), dtype=int))
+
+    return (
+        ns_results,
+        dream_results,
+        CalculationResults(x_data=[roughness, background, scalefactor], distribution=probability_array),
+    )
+
+
+def plot_posterior_comparison(
+    ns_results: RAT.outputs.BayesResults, dream_results: RAT.outputs.BayesResults, calc_results: CalculationResults
+):
+    """Create a grid of marginalised posteriors comparing different calculation methods.
+
+    Parameters
+    ----------
+    ns_results : RAT.BayesResults
+        The BayesResults object from a nested sampler calculation.
+    dream_results : RAT.BayesResults
+        The BayesResults object from a DREAM calculation.
+    calc_results : CalculationResults
+        The results from a direct calculation.
+    """
+    num_params = calc_results.distribution.ndim
+    fig, axes = plt.subplots(3, num_params, figsize=(3 * num_params, 9))
+
+    def plot_marginalised_result(dimension: int, axes: plt.Axes, limits: tuple[float]):
+        """Plot a histogram of a marginalised posterior from the calculation results.
+
+        Parameters
+        ----------
+        dimension : int
+            The dimension of the array to marginalise over.
+        axes : plt.Axes
+            The Axes object to plot the histogram onto.
+        limits : tuple[float]
+            The x-axis limits for the histogram.
+
+        """
+        # marginalise to the dimension
+        # note we don't need to normalise here as np.histogram normalises for us
+        sum_axes = tuple(i for i in range(0, num_params) if i != dimension)
+        distribution = np.sum(calc_results.distribution, axis=sum_axes)
+        distribution /= np.sum(calc_results.distribution)
+
+        # create histogram
+        axes.hist(
+            calc_results.x_data[i],
```

Comment:
The index `i` is not in scope here.

Reply:
weirdly, it kind of *was*: when `plot_marginalised_result` runs, the loop variable `i` is in the scope of `plot_posterior_comparison` and `plot_marginalised_result` is local inside of that scope! this should be `dimension` for good practice, but the plots don't actually change from it